libs/server/Resp/Vector/VectorManager.Replication.cs

```diff
@@ -0,0 +1,541 @@
+    /// </summary>
+    public sealed partial class VectorManager
+    {
+        /// <summary>
+        /// Represents a copy of a VADD being replayed during replication.
+        /// </summary>
+        private readonly record struct VADDReplicationState(Memory<byte> Key, uint Dims, uint ReduceDims, VectorValueType ValueType, Memory<byte> Values, Memory<byte> Element, VectorQuantType Quantizer, uint BuildExplorationFactor, Memory<byte> Attributes, uint NumLinks)
+        {
+        }
+
+        private int replicationReplayStarted;
+        private long replicationReplayPendingVAdds;
+        private readonly ManualResetEventSlim replicationBlockEvent;
+        private readonly Channel<VADDReplicationState> replicationReplayChannel;
+        private readonly Task[] replicationReplayTasks;
+
+        /// <summary>
+        /// For replication purposes, we need a write against the main log.
+        /// 
+        /// But we don't actually want to do the (expensive) vector ops as part of a write.
+        /// 
+        /// So this fakes up a modify operation that we can then intercept as part of replication.
+        /// 
+        /// This the Primary part, on a Replica <see cref="HandleVectorSetAddReplication"/> runs.
+        /// </summary>
+        internal void ReplicateVectorSetAdd<TContext>(ref SpanByte key, ref RawStringInput input, ref TContext context)
+            where TContext : ITsavoriteContext<SpanByte, SpanByte, RawStringInput, SpanByteAndMemory, long, MainSessionFunctions, MainStoreFunctions, MainStoreAllocator>
+        {
+            Debug.Assert(input.header.cmd == RespCommand.VADD, "Shouldn't be called with anything but VADD inputs");
+
+            var inputCopy = input;
+            inputCopy.arg1 = VADDAppendLogArg;
+
+            Span<byte> keyWithNamespaceBytes = stackalloc byte[key.Length + 1];
+            var keyWithNamespace = SpanByte.FromPinnedSpan(keyWithNamespaceBytes);
+            keyWithNamespace.MarkNamespace();
+            keyWithNamespace.SetNamespaceInPayload(0);
+            key.AsReadOnlySpan().CopyTo(keyWithNamespace.AsSpan());
+
+            Span<byte> dummyBytes = stackalloc byte[4];
+            var dummy = SpanByteAndMemory.FromPinnedSpan(dummyBytes);
+
+            var res = context.RMW(ref keyWithNamespace, ref inputCopy, ref dummy);
+
+            if (res.IsPending)
+            {
+                CompletePending(ref res, ref dummy, ref context);
+            }
+
+            if (!res.IsCompletedSuccessfully)
+            {
+                logger?.LogCritical("Failed to inject replication write for VADD into log, result was {res}", res);
+                throw new GarnetException("Couldn't synthesize Vector Set add operation for replication, data loss will occur");
+            }
+
+            // Helper to complete read/writes during vector set synthetic op goes async
+            static void CompletePending(ref Status status, ref SpanByteAndMemory output, ref TContext context)
```

Comment:
nit: looks like this function repeats 3 times on this file, should we have a common method?

Reply:
Consequence of these all being in different files and then consolidated into VectorManager over time.  Will DRY up.